
[![Generic badge](https://img.shields.io/badge/<journal>-<MPE>-<blue>.svg)](https://shields.io/)
[![DOI:10.1016/j.ympev.2025.108358](https://zenodo.org/badge/DOI/10.1016/j.ympev.2025.108358.svg)](https://doi.org/10.1016/j.ympev.2025.108358) 
[![Citation Badge](https://api.juleskreuer.eu/citation-badge.php?doi=10.1016/j.ympev.2025.108358)](https://juleskreuer.eu/citation-badge/)

# Squamata

`Repo for scripts and pipelines`

**Title:** Evidence of convergent evolution in the nuclear and mitochondrial OXPHOS genes subunits across the deep lineages of Squamata

**Authors:** Oscar Wallnoefer, Alessandro Formaggioni, Federico Plazzi, Marco Passamonti 

**Summary:** The order Squamata is a good candidate for detecting unusual patterns of mitochondrial evolution. The lineages leading to the snake and agamid clades likely experienced convergent evolution in mitochondrial OXidative PHOSphorylation (OXPHOS) genes, which provides strong support for the sister relationship of these two groups.

The OXPHOS subunits are encoded by both the nuclear and mitochondrial genomes, which are subject to distinct evolutionary pressures. Nevertheless, the cooperation between OXPHOS subunits is essential for proper OXPHOS function, as incompatibilities between subunits can be highly deleterious.
In the present study, we annotated OXPHOS genes of 56 Squamata species. The nuclear OXPHOS subunits that physically interact with mitochondrial proteins also support the clade sister relationship between snakes and agamids. Additionally, we found a significant number of convergent amino acid changes between agamids and snakes, not only in mitochondrial OXPHOS genes, but also in nuclear ones, with a higher rate of convergence in the nuclear OXPHOS subunits that play central roles in the OXPHOS complexes, like COX4 and NDUFA4.

Overall, the common selective pressures in two distinct lineages can lead two sets of genes, encoded by two different genomes, to exhibit similar patterns of convergent evolution, as well as similar evolutionary rates. As a consequence, the coevolution of interdependent subunits and their adaptation to specific evolutionary pressures can heavily influence the molecular structure of cytonuclear enzyme complexes and blur phylogenetic signals.

![GraphicalAbstract_2.png](GraphicalAbstract_2.png)


